src/tools/simeng/gdb/GDBStub.cc

```diff
@@ -0,0 +1,331 @@
+    std::string bufferString = buffer;
+
+    if (regex_match(bufferString, ack_regex)) {
+      if (verbose_)
+        std::cout << CYAN << "<- Received message acknowledgement" << std::endl
+                  << RESET;
+    } else if (regex_match(bufferString, packet_match, packet_regex)) {
+      std::string packet = packet_match[1].str();
+      if (verbose_) {
+        std::cout << CYAN;
+        if (packet_match.size() == 3)
+          std::cout << "<- Received packet: " << packet << std::endl;
+        else
+          std::cout << RED << "<- Unknown message"
+                    << std::endl;  // TODO: take action
+        std::cout << RESET;
+      }
+      std::string checksum = packet_match[2].str();
+      std::string expected = computeChecksum(packet);
+      if (verbose_ && checksum != expected) {
+        std::cout << RED << "   Checksum failed\n"
+                  << "   Expected checksum: " << expected
+                  << "   , but received: " << checksum << std::endl
+                  << RESET;
+        // TODO: take action
+      }
+
+      std::regex qSupported_regex("^qSupported:(.*)");
+      std::smatch qSupported_match;
+
+      std::string response = "+";  // acknowledgement
+
+      if (packet == "?") {  // reason for halting
+        response += generateReply("S05");
+      } else if (packet == "s") {  // step one instruction
+        core_.tick();
+        dataMemory_.tick();
+        response += generateReply("S05");
+      } else if (packet == "c") {  // continue until next breakpoint
+        std::string message = handleContinue();
+        response += generateReply(message);
+      } else if (packet == "g") {  // read registers
+        std::string registers = handleReadRegisters();
+        response += generateReply(registers);
+      } else if (packet[0] == 'p') {  // read single register
+        std::regex reg_regex("^p([0-9a-f]*)");
+        std::smatch reg_match;
+        regex_match(packet, reg_match, reg_regex);
+        std::string registerValue = handleReadSingleRegister(reg_match[1]);
+        response += generateReply(registerValue);
+      } else if (packet[0] == 'm') {  // read memory
+        std::regex mem_regex("^m([0-9a-f]*),([0-9a-f]*)");
+        std::smatch mem_match;
+        regex_match(packet, mem_match, mem_regex);
+        std::string memoryValue = handleReadMemory(mem_match[1], mem_match[2]);
+        response += generateReply(memoryValue);
+      } else if (packet[0] == 'Z' ||
+                 packet[0] == 'z') {  // create/remove breakpoint
+        std::regex break_regex(
+            "^([zZ])([0-4]),([0-9a-f]+),([0-9a-f]+)");  // format:
+                                                        // z/Z,type,address,kind
+        std::smatch break_match;
+        regex_match(packet, break_match, break_regex);
+        std::string type = break_match[2];
+        std::string address = break_match[3];
+        // std::string kind = break_match[4]; // kind not used
```

Comment:
If un-used then could remove this comment entirely, unless it provides some benefit to currently un-supported features?

Reply:
it looks like breakpoint kinds could be used for thumb mode - is the line worth keeping in for usefulness later, perhaps with a thumb mention ?

Reply:
Keep the comment, but it may be useful to include what it can be used for in the handover docs

Reply:
thanks, will do